doc: Add (legacy) requirements document for Device SDKs (#190)

iain-anderson opened the pull request:
Signed-off-by: Iain Anderson <iain@iotechsys.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x". -->

- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->

Some aspects of device service functional requirements are defined in a google document linked from the [WG wiki](https://wiki.edgexfoundry.org/display/FA/Device+and+Device+SDK+Working+Group)

Issue Number: #190 

## What has been updated?
Markdown document added to legacy requirements folder

## Are there any specific instructions or things that should be known prior to reviewing?

## Other information

jpwhitemn commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,153 @@
+configuration is held in the `Driver` section of the configuration file or
+registry.
+
+The SDK must also supply a Logging client at this stage.
```
Is this true anymore with logging optional?

iain-anderson replied:
Perhaps 'facility' would be better than 'client'? The facility should emit logs locally by default but should instead use the optional logging service if so configured.

jpwhitemn commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,153 @@
+
+The SDK should also handle environment variables. In addition to the common
+EdgeX variables, `EDGEX_INSTANCE_NAME` should if set override the `--instance`
+setting.
```
In accordance with https://docs.edgexfoundry.org/1.2/design/adr/0005-Service-Self-Config/

iain-anderson replied:
updated

lenny-intel commented on docs_src/design/legacy-requirements/README.md:
```diff
@@ -1 +1,8 @@
+
+| Name/Link                                          | Short Description                            |
+| -------------------------------------------------- | -------------------------------------------- |
+| [Device Service](device-service.md)                | Device Service SDK required functionality    |
```
This should be the TOC.md and each doc have a leading 3 digit number?
Same as this: https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/design/TOC.md

iain-anderson replied:
I was taking the lead from https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/design/legacy-design/README.md

iain-anderson commented:
> @iain-anderson - this is a good start. It seems we are missing elements from here https://wiki.edgexfoundry.org/display/FA/Device+and+Device+SDK+Working+Group?preview=/329488/46760195/edgex-device-service-requirements-v11.pdf such as provision watcher, health checks, sys mgt support, binary support, and acuation support to name some. Is this document meant to be a replacement for the old requirement documents?

It's not intended to be a full replacement - it should cover everything that is not dealt with elsewhere. So the requirements covered in the linked docs for the REST API and the Discovery mechanism are omitted, as are elements common to all EdgeX services.

cloudxxx8 commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,157 @@
+
+Each device may have as part of its definition in Metadata a number of `AutoEvents` associated with it. An `AutoEvent` has the following fields:
+
+* **resource**: the name of a deviceresource or resourcecommand indicating what to read.
```
we renamed resourcecommand to device command

iain-anderson replied:
Yes. Updated, thanks

jpwhitemn reviewed (dismissed):
LTGM

tonyespy commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,157 @@
+* `Name` - the name of the device service
+* `Description` - an optional brief description of the service
+* `Labels` - optional string labels
+* `BaseAddress` - URL of the base of the service's REST API
```
Does it include a **scheme** (e.g. http://edgex-core-data:480080) or is it just **Host:Port**?

lenny-intel replied:
Yes, it is a fully qualified URL

tonyespy commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,157 @@
+Discovery/Interval | Int | Time between automatic discovery runs, in seconds. Defaults to zero (do not run discovery automatically).
+MaxCmdOps | Int | Defines the maximum number of resource operations that can be sent to the driver in a single command.
+MaxCmdResultLen | Int | Maximum string length for command results returned from the driver.
+UpdateLastConnected | Bool | If true, update the LastConnected attribute of a device whenever it is successfully accessed. Defaults to false.
```
Is the plan for 2.0/Ireland that we'd remove the code in Core Data that does this or are we deprecating the behavior in the SDKs? It's not clear to me whether there ever was an actual decision on this. I checked both device-sdk-go and edgex-go and couldn't find an open issue for removal of the logic from either...

lenny-intel commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,157 @@
+ConnectRetries | Int | Number of times to attempt to contact core-data and core-metadata when starting up.
+StartupMsg | String | Message to log on successful startup.
+CheckInterval | String | The checking interval to request if registering with Consul
+ServerBindAddr | String | The interface on which the service's REST server should listen. By default the server listens on all available interfaces.
```
If ServerBindAddr is left blank it defaults to the Host value.
https://github.com/edgexfoundry/go-mod-bootstrap/blob/master/bootstrap/handlers/httpserver/httpserver.go#L85

lenny-intel commented on docs_src/design/legacy-requirements/device-service.md:
```diff
@@ -0,0 +1,157 @@
+Timeout | Int | Time (in milliseconds) to wait between attempts to contact core-data and core-metadata when starting up.
+ConnectRetries | Int | Number of times to attempt to contact core-data and core-metadata when starting up.
+StartupMsg | String | Message to log on successful startup.
+CheckInterval | String | The checking interval to request if registering with Consul
```
Correct, it is the interval we tell consul to use when sending the ping health check.

iain-anderson commented:
Rebased and updated the entry for `ServerBindAddress` to indicate
1) Default should be to use the `Host` value
2) `0.0.0.0` indicates all interfaces

iain-anderson commented:
rebased again @tonyespy @lenny-intel

lenny-intel commented:
I am deferring to @tonyespy  on this review.

tonyespy reviewed (approved):
LGTM

iain-anderson commented:
> I am deferring to @tonyespy on this review.

@lenny-intel  you still need to clear your request for changes before this can be merged

jpwhitemn commented:
LGTM

cloudxxx8 reviewed (approved):
LGTM
